foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/asm/ASMFactory.java

```diff
@@ -0,0 +1,361 @@
+        }
+    }
+
+    public static Type createType(final String typeDescriptor) {
+        String asmService = ASMFactory.getAsmService();
+        if (ASM_SERVICE_ECLIPSELINK.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.eclipselink.TypeImpl(typeDescriptor);
+        } else if (ASM_SERVICE_OW2.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.ow2.TypeImpl(typeDescriptor);
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static Type createVoidType() {
+        String asmService = ASMFactory.getAsmService();
+        if (ASM_SERVICE_ECLIPSELINK.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.eclipselink.TypeImpl((Class<?>)null);
+        } else if (ASM_SERVICE_OW2.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.ow2.TypeImpl((Class<?>)null);
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static Label createLabel() {
+        String asmService = ASMFactory.getAsmService();
+        if (ASM_SERVICE_ECLIPSELINK.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.eclipselink.LabelImpl();
+        } else if (ASM_SERVICE_OW2.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.ow2.LabelImpl();
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static SerialVersionUIDAdder createSerialVersionUIDAdder(final ClassVisitor classVisitor) {
+        String asmService = ASMFactory.getAsmService();
+        if (ASM_SERVICE_ECLIPSELINK.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.eclipselink.SerialVersionUIDAdderImpl(classVisitor);
+        } else if (ASM_SERVICE_OW2.equals(asmService)) {
+            return new org.eclipse.persistence.asm.internal.platform.ow2.SerialVersionUIDAdderImpl(classVisitor);
+        } else {
+            throw ValidationException.notAvailableASMService();
+        }
+    }
+
+    public static String getAsmService() {
+        String asmService = PrivilegedAccessHelper.getSystemProperty(SystemProperties.ASM_SERVICE);
+        if (asmService != null) {
+            if (ASM_SERVICE_ECLIPSELINK.equals(asmService) && isASMImplementationAvailable(ASM_ECLIPSELINK_CLASS_VISITOR)) {
+                LOG.finest("EclipseLink ASM implementation is used.");
+                return asmService;
+            } else if (ASM_SERVICE_OW2.equals(asmService) && isASMImplementationAvailable(ASM_OW2_CLASS_VISITOR)) {
+                LOG.finest("OW2 ASM implementation is used.");
+                return asmService;
+            } else {
+                throw ValidationException.incorrectASMServiceProvided();
```

Comment:
Not having access to ASM is kind of a big deal, as without it features like lazy loading go out the window.  I think it might be better to emit a warning here that the configuration option was bogus, and then levy the fallback behavior.